CatalogReaderExample/NuGet.Protocol.Catalog/CatalogProcessor.cs

```diff
@@ -0,0 +1,214 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using NuGet.Protocol.Core.Types;
+
+namespace NuGet.Protocol.Catalog
+{
+    public class CatalogProcessor
+    {
+        private const string CatalogResourceType = "Catalog/3.0.0";
+        private readonly ICatalogLeafProcessor _leafProcessor;
+        private readonly ICatalogClient _client;
+        private readonly ICursor _cursor;
+        private readonly ILogger<CatalogProcessor> _logger;
+        private readonly CatalogProcessorSettings _settings;
+
+        public CatalogProcessor(
+            ICursor cursor,
+            ICatalogClient client,
+            ICatalogLeafProcessor leafProcessor,
+            CatalogProcessorSettings settings,
+            ILogger<CatalogProcessor> logger)
+        {
+            _leafProcessor = leafProcessor ?? throw new ArgumentNullException(nameof(leafProcessor));
+            _client = client ?? throw new ArgumentNullException(nameof(client));
+            _cursor = cursor ?? throw new ArgumentNullException(nameof(cursor));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+
+            if (settings == null)
+            {
+                throw new ArgumentNullException(nameof(settings));
+            }
+
+            if (settings.ServiceIndexUrl == null)
+            {
+                throw new ArgumentException(
+                    $"The {nameof(CatalogProcessorSettings.ServiceIndexUrl)} property of the " +
+                    $"{nameof(CatalogProcessorSettings)} must not be null.");
+            }
+
+            // Clone the settings to avoid mutability issues.
+            _settings = new CatalogProcessorSettings
```

Comment:
add a Clone method to CatalogProcessorSettings

Reply:
Good idea. I will make it `internal` for now.

Reply:
Fixed.